src/compiler/checker.ts

```diff
@@ -6508,13 +6508,18 @@ namespace ts {
                 let hasRestLikeParameter = hasRestParameter(declaration);
                 if (!hasRestLikeParameter && isInJavaScriptFile(declaration)) {
                     const lastParam = lastOrUndefined(declaration.parameters);
-                    const lastParamIsDocumentedAsRest = !!lastParam && some(getJSDocParameterTags(lastParam), p =>
-                        p.typeExpression && p.typeExpression.type.kind === SyntaxKind.JSDocVariadicType);
-                    if (lastParamIsDocumentedAsRest || containsArgumentsReference(declaration)) {
+                    const lastParamTags = lastParam && getJSDocParameterTags(lastParam);
+                    const lastParamVariadicType = lastParamTags && firstDefined(lastParamTags, p =>
+                        p.typeExpression && isJSDocVariadicType(p.typeExpression.type) ? p.typeExpression.type : undefined);
+                    if (lastParamVariadicType || containsArgumentsReference(declaration)) {
                         hasRestLikeParameter = true;
                         const syntheticArgsSymbol = createSymbol(SymbolFlags.Variable, "args" as __String);
-                        syntheticArgsSymbol.type = lastParamIsDocumentedAsRest ? createArrayType(getTypeOfParameter(lastParam.symbol)) : anyArrayType;
+                        syntheticArgsSymbol.type = lastParamVariadicType ? createArrayType(getTypeFromTypeNode(lastParamVariadicType.type)) : anyArrayType;
                         syntheticArgsSymbol.isRestParameter = true;
+                        if (lastParamVariadicType) {

```

Comment:
Just thinking aloud: About half of this code has a separate code path for `lastParamVariadicType`. Would it read better if the normal and last-param-variadic parts were separate?

Reply:
Moved this whole part out to its own function.